### PR TITLE
TE-7923 Adjusted SprykerInstall module to be extendable

### PR DESCRIPTION
**Pull request by asaulenko**
- Developer(s): @asaulenko

- Ticket: https://spryker.atlassian.net/browse/TE-7923

- Academy PR: ACADEMY_URL_HERE

- Release Group: https://release.spryker.com/release-groups/view/3173

#### Documentation

- [ ] Functional documentation provided or in progress?
- [ ] Integration guide for projects provided or not needed?
- [ ] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   Install               | major  |                       |

-----------------------------------------

##### Change log

Breaking Changes

- Adjusted package classes to turn it into Spryker module with the aim of making it extendable on the project level.
- `\Spryker\Console\InstallConsoleCommand` was removed in the favor of  `\Spryker\Zed\Install\Communication\Console\InstallConsole`. Binary file (vendor/bin/install) not uses a new one.
- Infrastructure classes (`InstallFacade`, `InstallBusinessFactory`, `InstallConfig`, etc.) now can be extended on the project level. Original classes were removed.

### CODE REVIEW

**Review by dereuromark — CHANGES_REQUESTED** [review]
This wasnt a module so far, but a tool
If you now make this one, this requires some major changes also in tooling and release app

https://github.com/spryker/install/blob/master/composer.json#L3
And a few more important things like Spryker CI as requirement before merging.

Also, I need to fix the release app to import the existing releases as well as change the type from library to module etc.

File: src/Spryker/Zed/Install/Business/Runner/Section/SectionRunnerInterface.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+interface SectionRunnerInterface
+{
+    /**
+     * @param \Spryker\Zed\Install\Business\Stage\Section\SectionInterface $section
+     * @param \Spryker\Zed\Install\Business\Configuration\ConfigurationInterface $configuration
+     *
+     * @return void
+     */
+    public function run(SectionInterface $section, ConfigurationInterface $configuration);
```
Comment (stereomon) [inline src/Spryker/Zed/Install/Business/Runner/Section/SectionRunnerInterface.php:21]: Return type is missing.

File: bin/install [hunk 2]
```diff
@@ -1,10 +1,16 @@
 #!/usr/bin/env php
 <?php
 
+define('APPLICATION_ENV', (getenv('APPLICATION_ENV') ?: 'development'));
```
Comment (stereomon) [inline bin/install:4]: I thought you will use the Environment class? That would already have probably all you need. Or was there something blocking you from using it?

**Review by stereomon — CHANGES_REQUESTED** [review]

File: bin/install (hunk 2 above)
Comment (asaulenko) [reply bin/install:4]: I am confused. Haven’t we agreed on using the approach with manually define ing the constants?
But yes, there is a problem actually.
Environment class itself required the `APPLICATION` constant to be defined.
And it’s mandatory.

File: src/Spryker/Zed/Install/Business/Configuration/ConfigurationInterface.php [hunk 3]
```diff
@@ -5,35 +5,35 @@
...
 
-namespace Spryker\Install\Configuration;
+namespace Spryker\Zed\Install\Business\Configuration;
 
-use Spryker\Install\Stage\Section\Command\CommandInterface;
-use Spryker\Install\Stage\StageInterface;
-use Spryker\Style\StyleInterface;
+use Spryker\Zed\Install\Business\Stage\Section\Command\CommandInterface;
+use Spryker\Zed\Install\Business\Stage\StageInterface;
+use Spryker\Zed\Install\Communication\Style\StyleInterface;
```
Comment (gechetspr) [inline src/Spryker/Zed/Install/Business/Configuration/ConfigurationInterface.php:12]: Mb we can put `StyleInterface` to the Shared layer in order to not mix business and communication layers?

File: src/Spryker/Zed/Install/Business/Executable/ExecutableFactory.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    /**
+     * @param \Spryker\Zed\Install\Business\Stage\Section\Command\CommandInterface $command
+     * @param \Spryker\Zed\Install\Business\Configuration\ConfigurationInterface $configuration
+     *
+     * @return \Spryker\Zed\Install\Business\Executable\CommandLine\CommandLineExecutable
+     */
+    protected function createCommandLineExecutable(
+        CommandInterface $command,
+        ConfigurationInterface $configuration
+    ): CommandLineExecutable {
```
Comment (gechetspr) [inline src/Spryker/Zed/Install/Business/Executable/ExecutableFactory.php:38]: why not an interface as a return type?

File: src/Spryker/Zed/Install/Business/Executable/ExecutableFactory.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace Spryker\Zed\Install\Business\Executable;
+
+use Spryker\Zed\Install\Business\Configuration\ConfigurationInterface;
+use Spryker\Zed\Install\Business\Executable\CommandLine\CommandLineExecutable;
+use Spryker\Zed\Install\Business\Stage\Section\Command\CommandInterface;
+
+class ExecutableFactory
```
Comment (gechetspr) [inline src/Spryker/Zed/Install/Business/Executable/ExecutableFactory.php:14]: Mb we don't need additional factories and can put everything in a one?

File: src/Spryker/Zed/Install/Business/InstallFacade.php [hunk 6]
```diff
@@ -0,0 +1,38 @@
+<?php
+
+/**
+ * MIT License
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace Spryker\Zed\Install\Business;
+
+use Spryker\Zed\Install\Communication\CommandLine\CommandLineArgumentContainer;
```
Comment (gechetspr) [inline src/Spryker/Zed/Install/Business/InstallFacade.php:10]: Same. Mb we can move those classes to shared layer if we also need them in the communication one?

**Review by gechetspr — CHANGES_REQUESTED** [review]
Only few changes required

File: src/Spryker/Zed/Install/Business/Configuration/ConfigurationInterface.php (hunk 3 above)
Comment (asaulenko) [reply src/Spryker/Zed/Install/Business/Configuration/ConfigurationInterface.php:12]: Well, we kinda can. But afaik `Shared` layer is for stuff that's meant to be used across different applications (`Yves`, `Zed`, `Client` etc), not across the layers of `Zed` application.
@stereomon what do you say?

File: src/Spryker/Zed/Install/Business/Executable/ExecutableFactory.php (hunk 5 above)
Comment (asaulenko) [reply src/Spryker/Zed/Install/Business/Executable/ExecutableFactory.php:14]: I don't see an easy way to do that. This factory is parametrized and is used to process dynamic values of `$command`, `$configuration` into `$executable` inside the `CommandRunner` model. Extracting it from `CommandRunnder` implies heavy refactoring of this class and most probably all the related tests.

File: src/Spryker/Zed/Install/Business/Configuration/ConfigurationInterface.php (hunk 3 above)
Comment (stereomon) [reply src/Spryker/Zed/Install/Business/Configuration/ConfigurationInterface.php:12]: I think it's a good idea to move that to Shared.

**Review by gechetspr — APPROVED** [review]

File: src/Spryker/Zed/Install/Business/InstallFacadeInterface.php [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+namespace Spryker\Zed\Install\Business;
+
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+interface InstallFacadeInterface
+{
+    /**
+     * Specification:
+     * - Triggers the installation process.
```
Comment (gechetspr) [inline src/Spryker/Zed/Install/Business/InstallFacadeInterface.php:17]:
```suggestion
     * - Triggers an installation process.
```

I think this is more correct as long as install process can differ from env to env

File: README.md [hunk 8]
```diff
@@ -1,5 +1,4 @@
-# Spryker Install
-
+# Install Module
```
Comment (gechetspr) [inline README.md:1]: Please adjust the text in this readme too, please

File: src/Spryker/Zed/Install/Communication/Console/InstallConsole.php [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+     * @param \Symfony\Component\Console\Input\InputInterface $input
+     * @param \Symfony\Component\Console\Output\OutputInterface $output
+     *
+     * @return int
+     */
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $this->getFacade()->runInstall($input, $output);
+
+        return 0;
```
Comment (gechetspr) [inline src/Spryker/Zed/Install/Communication/Console/InstallConsole.php:86]: Please move this to the constant, e.g. EXIT_CODE_SUCCESS

**Review by gechetspr — CHANGES_REQUESTED** [review]
Minor changes required.

Also please adjust changelogs and release type

File: src/Spryker/Zed/Install/Business/CommandLine/CommandLineOptionContainer.php [hunk 10]
```diff
@@ -96,31 +96,31 @@ public function getRecipe(): string
...
     public function getExclude(): array
     {
         return $this->exclude;
     }
 
     /**
-     * @return array
+     * @return string[]
      */
     public function getIncludeExcluded(): array
```
[the remark this answers is not in the document]
Comment (asaulenko) [reply src/Spryker/Zed/Install/Business/CommandLine/CommandLineOptionContainer.php:125]: `Include command(s)/section(s) which are marked as excluded in the configuration.`
I was not able to pick a more descriptive name for it.

**Review by gechetspr — APPROVED** [review]

**Comment by gechetspr** [comment]
Major approved by @PhilinTv

**Review by dereuromark — APPROVED** [review]

**Pull request merged**